Implement parsing for proxy protocol v2.

btimby opened the pull request:
Hey Carlo, I am back again with another addition. This adds support for the haproxy proxy protocol version 2. This version uses a binary format and can include additional fields. Additional fields are TLV (type, length, value) records that are optional. The project that I use `proxied-socket` in requires this new format. The proxy protocol is documented at the link below.

https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt

I included tests which mirror the existing tests for the haproxy and proxied-socket formats. I used an actual header from haproxy in the tests.

B3rn475 commented:
First of all, huge **Thank You** for all your contributions.

Small nit:
Could you rename `lib/proxy-v2.js` to lib/`haproxy-v2.js` for symmetry?

I'll take a deeper look to the pull request ASAP

B3rn475 commented on lib/server/header/proxy-v2.js:
```diff
@@ -0,0 +1,135 @@
+  };
+
+  let offset = 0;
+  if (header.family === 0x01) {
+    fields.remoteFamily = 'IPv4';
+    fields.remoteAddress = decodeV4Address(buffer, offset);
+    fields.localAddress = decodeV4Address(buffer, offset += 4);
+    fields.remotePort = buffer.readUInt16BE(offset += 4);
+    fields.localPort = buffer.readUInt16BE(offset += 2);
+    offset += 2;
+  } else if (header.family === 0x02) {
+    fields.remoteFamily = 'IPv6';
+    fields.remoteAddress = decodeV6Address(buffer, offset);
+    fields.localAddress = decodeV6Address(buffer, offset += 16);
+    fields.remotePort = buffer.readUInt16BE(offset += 16);
+    fields.localPort = buffer.readUInt16BE(offset += 2);
+    offset += 2;
+  }
```
Here and elsewhere in the file, could you add some extra error checking logic?

Here for example if the family is neither IPv4 nor IPv6 we should throw an Error

btimby replied:
Done. I also added support for protocol families `0` and `3`.

btimby commented:
Thank you for the suggestions, both were good. I renamed the module and added appropriate checks / throws. I was in a hurry to get this ready for my project, so I did not do much work once it was working :-).

In addition to passing tests, I can attest that this works in a real-world scenario, I am installing my branch in my project and have it reading the v2 header, including multiple TLV values (the motivation for this change was to allow passing TLV values to my services). Specifically, I want the SSL fields (I use SNI) and `Unique-ID` from haproxy so I can correlate logs within my entire stack.

Just to let you know, I looked at other libraries that handle proxy protocol v2, and I prefer how this wraps the socket / server. Plus I am already using this library, so this change is a drop-in for me. I found proxy protocol v2 parsers, but none supported extracting TLV, they just skipped them.

B3rn475 commented:
Thank you very much for the contribution.

If your project is public I would love to know about it.
We could even have a "Projects that use `proxied-socket`" section in the `README.md` file.

B3rn475 commented:
The provided changes, together with a bump in dependencies, has been published on npm as version `1.2.0`.
